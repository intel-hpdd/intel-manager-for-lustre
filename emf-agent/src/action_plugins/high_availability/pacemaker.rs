// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

/// The order of Role is order of preference for returning in a cloned
/// pacemaker resource.
#[derive(Debug, PartialEq, PartialOrd, Eq, Ord)]
pub enum Role {
    Error(String),
    Started,
    Starting,
    Stopping,
    Stopped,
}

impl From<&str> for Role {
    fn from(role: &str) -> Self {
        match role {
            "Started" => Role::Started,
            "Starting" => Role::Starting,
            "Stopping" => Role::Stopping,
            "Stopped" => Role::Stopped,
            other => Role::Error(other.to_string()),
        }
    }
}
