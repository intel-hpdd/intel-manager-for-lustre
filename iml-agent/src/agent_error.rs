// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use iml_cmd::CmdError;
use iml_fs::ImlFsError;
use iml_wire_types::PluginName;
use std::{fmt, process::Output};
use thiserror::Error;
use tokio_util::codec::LinesCodecError;

pub type Result<T> = std::result::Result<T, ImlAgentError>;

#[derive(Debug)]
pub struct NoSessionError(pub PluginName);

impl fmt::Display for NoSessionError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "No session found for {:?}", self.0)
    }
}

impl std::error::Error for NoSessionError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        None
    }
}

#[derive(Debug)]
pub struct NoPluginError(pub PluginName);

impl fmt::Display for NoPluginError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "Plugin in registry not found for {:?}", self.0)
    }
}

impl std::error::Error for NoPluginError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        None
    }
}

#[derive(Debug)]
pub struct RequiredError(pub String);

impl fmt::Display for RequiredError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}", self.0)
    }
}

impl std::error::Error for RequiredError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        None
    }
}

#[derive(Debug)]
pub struct CibError(pub String);

impl fmt::Display for CibError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl std::error::Error for CibError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        None
    }
}

#[derive(Debug, Error)]
pub enum ImlAgentError {
    #[error(transparent)]
    AddrParseError(#[from] std::net::AddrParseError),
    #[error(transparent)]
    IpNetworkError(#[from] ipnetwork::IpNetworkError),
    #[error(transparent)]
    CibError(#[from] CibError),
    #[error(transparent)]
    CmdError(#[from] CmdError),
    #[error(transparent)]
    FmtError(#[from] strfmt::FmtError),
    #[error(transparent)]
    FromUtf8Error(#[from] std::string::FromUtf8Error),
    #[error(transparent)]
    ImlFsError(#[from] ImlFsError),
    #[error(transparent)]
    InvalidHeaderValue(#[from] http::header::InvalidHeaderValue),
    #[error(transparent)]
    InvalidUri(#[from] http::uri::InvalidUri),
    #[error(transparent)]
    InvalidUriParts(#[from] http::uri::InvalidUriParts),
    #[error(transparent)]
    Io(#[from] std::io::Error),
    #[error(transparent)]
    LiblustreError(#[from] liblustreapi::error::LiblustreError),
    #[error(transparent)]
    LinesCodecError(#[from] LinesCodecError),
    #[error(transparent)]
    LustreCollectorError(#[from] lustre_collector::error::LustreCollectorError),
    #[error("Marker Not Found")]
    MarkerNotFound,
    #[error("Argument '{0}' Missing")]
    MissingArgument(String),
    #[error(transparent)]
    NoPluginError(#[from] NoPluginError),
    #[error(transparent)]
    NoSessionError(#[from] NoSessionError),
    #[error(transparent)]
    OneshotCanceled(#[from] futures::channel::oneshot::Canceled),
    #[error(transparent)]
    ParseBoolError(#[from] std::str::ParseBoolError),
    #[error(transparent)]
    ParseIntError(#[from] std::num::ParseIntError),
    #[error(transparent)]
    RequiredError(#[from] RequiredError),
    #[error(transparent)]
    Reqwest(#[from] reqwest::Error),
    #[error("Rx went away")]
    SendError,
    #[error(transparent)]
    Serde(#[from] serde_json::Error),
    #[error(transparent)]
    SystemdError(#[from] iml_systemd::SystemdError),
    #[error(transparent)]
    TokioJoinError(#[from] tokio::task::JoinError),
    #[error(transparent)]
    TokioTimerError(#[from] tokio::time::Error),
    #[error("Unexpected status code")]
    UnexpectedStatusError,
    #[error(transparent)]
    UrlParseError(#[from] url::ParseError),
    #[error(transparent)]
    Utf8Error(#[from] std::str::Utf8Error),
    #[error(transparent)]
    XmlError(#[from] elementtree::Error),
    #[error(transparent)]
    QuickXmlError(#[from] quick_xml::Error),
    #[error("{0}")]
    LdevEntriesError(String),
    #[error(transparent)]
    CombineEasyError(#[from] combine::stream::easy::Errors<char, String, usize>),
}

impl From<Output> for ImlAgentError {
    fn from(output: Output) -> Self {
        ImlAgentError::CmdError(output.into())
    }
}

impl From<futures::channel::mpsc::SendError> for ImlAgentError {
    fn from(_: futures::channel::mpsc::SendError) -> Self {
        ImlAgentError::SendError
    }
}

impl serde::Serialize for ImlAgentError {
    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&format!("{:?}", self))
    }
}
