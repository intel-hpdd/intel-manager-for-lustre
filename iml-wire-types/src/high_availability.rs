// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use std::{
    convert::TryFrom,
    fmt::{self, Display},
    io,
};

#[derive(Debug, serde::Serialize, serde::Deserialize)]
pub enum NodeType {
    Unknown,
    Member,
    Remote,
    Ping,
}

impl TryFrom<&str> for NodeType {
    type Error = io::Error;

    fn try_from(x: &str) -> Result<Self, Self::Error> {
        match x {
            "unknown" => Ok(Self::Unknown),
            "member" => Ok(Self::Member),
            "remote" => Ok(Self::Remote),
            "ping" => Ok(Self::Ping),
            x => Err(io::Error::new(
                io::ErrorKind::InvalidInput,
                format!("Invalid node type {}", x),
            )),
        }
    }
}

impl Display for NodeType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let x = match self {
            NodeType::Unknown => "unknown",
            NodeType::Member => "member",
            NodeType::Remote => "remote",
            NodeType::Ping => "ping",
        };

        write!(f, "{}", x)
    }
}

#[derive(Debug, serde::Serialize, serde::Deserialize)]
pub struct Node {
    pub name: String,
    pub id: String,
    pub online: bool,
    pub standby: bool,
    pub standby_onfail: bool,
    pub maintenance: bool,
    pub pending: bool,
    pub unclean: bool,
    pub shutdown: bool,
    pub expected_up: bool,
    pub is_dc: bool,
    pub resources_running: u32,
    pub r#type: NodeType,
}

#[derive(Debug, Default, serde::Serialize, serde::Deserialize)]
pub struct Cluster {
    pub nodes: Vec<Node>,
}
