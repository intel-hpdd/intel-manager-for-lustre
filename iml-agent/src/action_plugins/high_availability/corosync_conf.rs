// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use crate::agent_error::ImlAgentError;
use iml_cmd::{CheckedCommandExt, Command};
use lazy_static::lazy_static;
use regex::Regex;
use std::{collections::HashMap, convert::TryFrom, io};
use tokio::{fs::OpenOptions, io::AsyncWriteExt};

static COROSYNC_CONF: &str = "/etc/corosync/corosync.conf";

/// Reads the corosync configuration.
async fn read_corosync_conf() -> Result<Vec<u8>, io::Error> {
    iml_fs::read_file_to_end(COROSYNC_CONF).await
}

/// Truncates and writes a new configuration.
async fn write_corosync_conf(xs: &[u8]) -> Result<(), ImlAgentError> {
    OpenOptions::new()
        .write(true)
        .truncate(true)
        .open(COROSYNC_CONF)
        .await?
        .write_all(xs)
        .await?;

    Ok(())
}

/// Updates the config_version. This is useful so nodes can tell which conf is newer.
fn update_config_version(
    s: impl Iterator<Item = String>,
    next_version: u64,
) -> impl Iterator<Item = String> {
    s.filter(|x| !x.trim().starts_with("config_version:"))
        .map(move |x| {
            if x.trim().starts_with("version:") {
                format!("{}\n    config_version: {}", x, next_version)
            } else {
                x
            }
        })
}

/// Updates the mcastport for all interfaces in the corosync.conf.
fn update_mcast_port(
    s: impl Iterator<Item = String>,
    new_port: u16,
) -> impl Iterator<Item = String> {
    lazy_static! {
        static ref RE: Regex = Regex::new(r"(\s*mcastport:\s*)\d+").unwrap();
    }

    s.map(move |x| {
        if x.trim().starts_with("mcastport:") {
            RE.replace(&x, |caps: &regex::Captures| {
                format!("{}{}", &caps[1], new_port)
            })
            .to_string()
        } else {
            x
        }
    })
}

/// Given a new mcast port, updates the port on each interface,
/// bumps the config_version and writes out the new corosync.conf.
pub(crate) async fn change_mcast_port(new_port: u16) -> Result<(), ImlAgentError> {
    let conf = read_corosync_conf().await?;
    let conf = std::str::from_utf8(&conf)?
        .trim()
        .lines()
        .map(|x| x.to_string());

    let (totem, _) = corosync_cmapctl().await?;

    let conf = update_config_version(conf, totem.next_version());
    let conf = update_mcast_port(conf, new_port);
    let conf = conf.collect::<Vec<_>>().join("\n");

    write_corosync_conf(conf.as_bytes()).await?;

    Ok(())
}

type StringMap<'a> = HashMap<&'a str, &'a str>;

#[derive(Debug)]
pub(crate) struct Totem {
    config_version: Option<u64>,
    cluster_name: String,
    token: u32,
}

impl Totem {
    fn next_version(&self) -> u64 {
        match self.config_version {
            Some(x) => x + 1,
            None => 1,
        }
    }
}

impl<'a> TryFrom<&'a StringMap<'a>> for Totem {
    type Error = ImlAgentError;

    fn try_from(x: &'a StringMap<'a>) -> Result<Self, Self::Error> {
        Ok(Totem {
            config_version: x
                .get("config_version (u64)")
                .map(|x| x.parse())
                .transpose()?,
            cluster_name: (*try_get(x, "cluster_name (str)")?).to_string(),
            token: try_get(x, "token (u32)")?.parse()?,
        })
    }
}

#[derive(Debug, Eq, PartialEq, Ord, PartialOrd)]
pub(crate) struct Interface {
    ringnumber: u32,
    bindnetaddr: String,
    mcastaddr: Option<String>,
    broadcast: Option<String>,
    mcastport: u16,
}

impl<'a> TryFrom<(&'a str, StringMap<'a>)> for Interface {
    type Error = ImlAgentError;

    fn try_from((ringnumber, x): (&'a str, StringMap<'a>)) -> Result<Self, Self::Error> {
        Ok(Interface {
            ringnumber: ringnumber.parse()?,
            bindnetaddr: (*try_get(&x, "bindnetaddr (str)")?).to_string(),
            mcastaddr: x.get("mcastaddr (str)").map(|x| (*x).to_string()),
            broadcast: x.get("broadcast (str)").map(|x| (*x).to_string()),
            mcastport: try_get(&x, "mcastport (u16)")?.parse()?,
        })
    }
}

/// View the current totem and interface information from `corosync-cmapctl`.
async fn corosync_cmapctl() -> Result<(Totem, Vec<Interface>), ImlAgentError> {
    let x = Command::new("corosync-cmapctl")
        .arg("totem")
        .kill_on_drop(true)
        .checked_output()
        .await?;

    corosync_cmapctl_parser(&x.stdout)
}

fn corosync_cmapctl_parser(x: &[u8]) -> Result<(Totem, Vec<Interface>), ImlAgentError> {
    let x = std::str::from_utf8(x)?;

    let (totem, interfaces) = x
        .trim()
        .lines()
        .filter_map(|y| {
            let xs: Vec<_> = y.split('=').collect();

            match xs[..] {
                [k, v] => Some((k.trim(), v.trim())),
                _ => None,
            }
        })
        .map(|(k, v)| (k.split('.').collect(), v))
        .fold(
            (HashMap::new(), HashMap::new()),
            |(mut totem, mut interfaces): (StringMap, HashMap<&str, StringMap>),
             (k, v): (Vec<_>, _)| {
                match k[..] {
                    ["totem", y] => {
                        totem.insert(y, v);
                    }
                    ["totem", "interface", x, y] => {
                        interfaces.entry(x).or_default().insert(y, v);
                    }

                    _ => {}
                }

                (totem, interfaces)
            },
        );

    Ok((
        Totem::try_from(&totem)?,
        interfaces
            .into_iter()
            .map(Interface::try_from)
            .collect::<Result<_, _>>()?,
    ))
}

fn try_get<'a>(hm: &'a StringMap<'a>, k: &str) -> Result<&'a &'a str, ImlAgentError> {
    hm.get(k)
        .ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, format!("Property {} not found", k)))
        .map_err(|e| e.into())
}

#[cfg(test)]
mod tests {
    use super::{corosync_cmapctl_parser, update_config_version, update_mcast_port};
    use std::iter::FromIterator;

    fn corosync_cmapctl_fixture() -> &'static [u8] {
        include_bytes!("../fixtures/corosync-cmapctl-totem.txt")
    }

    fn corosync_conf_fixture() -> &'static [u8] {
        include_bytes!("../fixtures/corosync.conf")
    }

    #[test]
    fn test_corosync_cmapctl_output() -> Result<(), Box<dyn std::error::Error>> {
        let x = corosync_cmapctl_fixture();

        let mut y = corosync_cmapctl_parser(x)?;

        y.1.sort_unstable();

        insta::assert_debug_snapshot!(y);

        Ok(())
    }

    #[test]
    fn test_update_config_version() -> Result<(), Box<dyn std::error::Error>> {
        let conf = corosync_conf_fixture();

        let conf = std::str::from_utf8(&conf)?
            .trim()
            .lines()
            .map(|x| x.to_string());

        let conf = Vec::from_iter(update_config_version(conf, 10)).join("\n");

        insta::assert_display_snapshot!(conf);

        Ok(())
    }

    #[test]
    fn test_update_mcast_port() -> Result<(), Box<dyn std::error::Error>> {
        let conf = corosync_conf_fixture();

        let conf = std::str::from_utf8(&conf)?
            .trim()
            .lines()
            .map(|x| x.to_string());

        let conf = Vec::from_iter(update_mcast_port(conf, 40001)).join("\n");

        insta::assert_display_snapshot!(conf);

        Ok(())
    }
}
