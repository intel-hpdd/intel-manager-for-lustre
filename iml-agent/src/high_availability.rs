// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use crate::agent_error::ImlAgentError;
use futures::TryFutureExt;
use iml_cmd::{CheckedCommandExt, Command};
use iml_fs::file_exists;
use iml_wire_types::high_availability::{Ban, Cluster, Node, Resource};
use quick_xml::{
    events::{attributes::Attributes, Event},
    Reader,
};
use std::{collections::HashMap, convert::TryInto, ffi::OsStr, process::Output};

static CIBADMIN_PATH: &str = "/usr/sbin/cibadmin";

fn cibadmin_cmd() -> Command {
    let mut cmd = Command::new(CIBADMIN_PATH);

    cmd.kill_on_drop(true);

    cmd
}

static CRM_MON_PATH: &str = "/usr/sbin/crm_mon";

pub(crate) fn crm_mon_cmd() -> Command {
    let mut cmd = Command::new(CRM_MON_PATH);

    cmd.arg("--one-shot")
        .arg("--inactive")
        .arg("--as-xml")
        .kill_on_drop(true);

    cmd
}

static CRM_RESOURCE_PATH: &str = "/usr/sbin/crm_resource";

fn crm_resource_cmd() -> Command {
    let mut cmd = Command::new(CRM_RESOURCE_PATH);

    cmd.kill_on_drop(true);

    cmd
}

static CRM_ATTRIBUTE_PATH: &str = "/usr/sbin/crm_attribute";

fn crm_attribute_cmd() -> Command {
    let mut cmd = Command::new(CRM_ATTRIBUTE_PATH);

    cmd.kill_on_drop(true);

    cmd
}

static PCS_PATH: &str = "/usr/sbin/pcs";

fn pcs_cmd() -> Command {
    let mut cmd = Command::new(PCS_PATH);

    cmd.kill_on_drop(true);

    cmd
}

static COROSYNC_CFGTOOL_PATH: &str = "/usr/sbin/corosync-cfgtool";

fn corosync_cfgtool() -> Command {
    let mut cmd = Command::new(COROSYNC_CFGTOOL_PATH);

    cmd.kill_on_drop(true);

    cmd
}

pub async fn get_local_nodeid() -> Result<Option<String>, ImlAgentError> {
    if !file_exists(COROSYNC_CFGTOOL_PATH).await {
        return Ok(None);
    }

    let output = corosync_cfgtool().arg("-s").checked_output().await?;

    parse_local_nodeid_output(&output.stdout)
}

fn parse_local_nodeid_output(x: &[u8]) -> Result<Option<String>, ImlAgentError> {
    let x = std::str::from_utf8(x)?;

    let prefix = "Local node ID ";

    let x = x
        .split_terminator('\n')
        .find(|x| x.starts_with(prefix))
        .and_then(|x| x.strip_prefix(prefix))
        .map(|x| x.to_string());

    Ok(x)
}

async fn get_lustre_resource_mounts(
    xs: Vec<String>,
) -> Result<HashMap<String, String>, ImlAgentError> {
    let mut out = HashMap::new();

    for x in xs {
        let output = crm_resource_cmd()
            .arg("--query-xml")
            .arg("--resource")
            .arg(&x)
            .checked_output()
            .await?;

        if let Some(v) = parse_resource_xml(&output.stdout)? {
            out.insert(x, v);
        };
    }

    Ok(out)
}

fn is_lustre_ra(x: &str) -> bool {
    x == "ocf::ddn:lustre-server" || x == "ocf::lustre:Lustre"
}

pub async fn set_resource_role(resource: &str, running: bool) -> Result<(), ImlAgentError> {
    crm_resource_cmd()
        .args(&[
            "--resource",
            resource,
            "--set-parameter",
            "target-role",
            "--meta",
            "--parameter-value",
            if running { "Started" } else { "Stopped" },
        ])
        .checked_output()
        .await?;

    Ok(())
}

pub async fn cibcreate(scope: &str, xml: &str) -> Result<(), ImlAgentError> {
    cibadmin_cmd()
        .args(&["--create", "--scope", scope, "--xml-text", xml])
        .checked_status()
        .inspect_err(|_| tracing::error!("Failed to create {}: {}", scope, xml))
        .err_into()
        .await
}

pub async fn cibxpath<I, S>(op: &str, xpath: &str, extra: I) -> Result<String, ImlAgentError>
where
    I: IntoIterator<Item = S>,
    S: AsRef<OsStr>,
{
    let o = cibadmin_cmd()
        .arg(format!("--{}", op))
        .arg("--xpath")
        .arg(xpath)
        .args(extra)
        .checked_output()
        .await?;

    Ok(String::from_utf8_lossy(&o.stdout).to_string())
}

pub async fn crm_resource<I, S>(args: I) -> Result<String, ImlAgentError>
where
    I: IntoIterator<Item = S>,
    S: AsRef<OsStr>,
{
    let o = crm_resource_cmd().args(args).checked_output().await?;

    Ok(String::from_utf8_lossy(&o.stdout).to_string())
}

pub async fn crm_attribute(args: Vec<String>) -> Result<String, ImlAgentError> {
    let o = crm_attribute_cmd().args(args).checked_output().await?;

    Ok(String::from_utf8_lossy(&o.stdout).to_string())
}

pub async fn pcs(args: Vec<String>) -> Result<String, ImlAgentError> {
    let o = pcs_cmd().args(args).checked_output().await?;

    Ok(String::from_utf8_lossy(&o.stdout).to_string())
}

pub async fn get_crm_mon_output() -> Result<Output, iml_cmd::CmdError> {
    crm_mon_cmd().checked_output().await
}

pub async fn get_crm_mon() -> Result<Option<Cluster>, ImlAgentError> {
    if !file_exists(CRM_MON_PATH).await {
        return Ok(None);
    }

    let crm_output = get_crm_mon_output().await?;

    let mut x = read_crm_output(&crm_output.stdout)?;

    let lustre_resource_ids = x
        .resources
        .iter()
        .filter_map(|x| {
            if is_lustre_ra(&x.resource_agent) {
                Some(x.id.to_string())
            } else {
                None
            }
        })
        .collect();

    let mounts = get_lustre_resource_mounts(lustre_resource_ids).await?;

    x.resource_mounts = mounts;

    let ban_output = crm_mon_cmd()
        .arg("--neg-locations")
        .checked_output()
        .await?;

    read_banned_output(&ban_output.stdout, &mut x)?;

    Ok(Some(x))
}

fn required_arg<'a>(arg: &str, x: &'a HashMap<&str, String>) -> Result<&'a str, ImlAgentError> {
    x.get(arg)
        .map(|x| x.as_str())
        .ok_or_else(|| ImlAgentError::MissingArgument(arg.into()))
}

fn required_bool(arg: &str, x: &HashMap<&str, String>) -> Result<bool, ImlAgentError> {
    let x = required_arg(arg, x)?;
    let x = x.parse()?;

    Ok(x)
}

fn node_from_map(x: &HashMap<&str, String>) -> Result<Node, ImlAgentError> {
    Ok(Node {
        name: required_arg("name", x)?.to_string(),
        id: required_arg("id", x)?.to_string(),
        online: required_bool("online", x)?,
        standby: required_bool("standby", x)?,
        standby_onfail: required_bool("standby_onfail", x)?,
        maintenance: required_bool("maintenance", x)?,
        pending: required_bool("pending", x)?,
        unclean: required_bool("unclean", x)?,
        shutdown: required_bool("shutdown", x)?,
        expected_up: required_bool("expected_up", x)?,
        is_dc: required_bool("is_dc", x)?,
        resources_running: required_arg("resources_running", x)?.parse::<u32>()?,
        r#type: required_arg("type", x)?.try_into()?,
    })
}

fn resource_from_map(x: &HashMap<&str, String>) -> Result<Resource, ImlAgentError> {
    Ok(Resource {
        id: required_arg("id", x)?.to_string(),
        resource_agent: required_arg("resource_agent", x)?.to_string(),
        role: required_arg("role", x)?.to_string(),
        active: required_bool("active", x)?,
        orphaned: required_bool("orphaned", x)?,
        managed: required_bool("managed", x)?,
        failed: required_bool("failed", x)?,
        failure_ignored: required_bool("failure_ignored", x)?,
        nodes_running_on: required_arg("nodes_running_on", x)?.parse::<u32>()?,
        active_node_id: None,
        active_node_name: None,
    })
}

fn ban_from_map(x: &HashMap<&str, String>) -> Result<Ban, ImlAgentError> {
    Ok(Ban {
        id: required_arg("id", x)?.to_string(),
        resource: required_arg("resource", x)?.to_string(),
        node: required_arg("node", x)?.to_string(),
        weight: required_arg("weight", x)?.parse::<i32>()?,
        master_only: required_bool("master_only", x)?,
    })
}

fn attrs_to_hashmap<'a>(
    mut attrs: Attributes<'a>,
    reader: &Reader<&[u8]>,
) -> Result<HashMap<&'a str, String>, ImlAgentError> {
    attrs.try_fold(HashMap::new(), |mut acc, x| {
        let x = x?;

        acc.insert(
            std::str::from_utf8(x.key)?,
            x.unescape_and_decode_value(reader)?,
        );

        Ok(acc)
    })
}

fn read_nodes(reader: &mut Reader<&[u8]>) -> Result<Vec<Node>, ImlAgentError> {
    let mut buf = vec![];
    let mut xs = vec![];

    loop {
        buf.clear();

        match reader.read_event(&mut buf)? {
            Event::Empty(x) if x.name() == b"node" => {
                let x = attrs_to_hashmap(x.attributes(), reader)?;
                let x = node_from_map(&x)?;

                xs.push(x);
            }
            Event::End(x) => {
                if x.name() == b"nodes" {
                    break;
                }
            }
            _ => {}
        }
    }

    Ok(xs)
}

enum Ra {
    Ddn(Option<String>),
    Lustre(Option<String>),
}

impl Ra {
    fn inner(self) -> Option<String> {
        match self {
            Self::Ddn(x) => x,
            Self::Lustre(x) => x,
        }
    }
}

fn parse_resource_xml(x: &[u8]) -> Result<Option<String>, ImlAgentError> {
    let x = std::str::from_utf8(x)?;

    let xs: Vec<_> = x.trim().lines().skip(2).collect();

    if xs.is_empty() {
        return Ok(None);
    }

    let x = xs.join("\n");
    let mut reader = Reader::from_str(&x);
    reader.trim_text(true);

    let mut buf = vec![];

    let mut current_ra = None;

    loop {
        match reader.read_event(&mut buf)? {
            Event::Start(ref x) if x.name() == b"primitive" => {
                let mut x = attrs_to_hashmap(x.attributes(), &reader)?;

                match (
                    x.remove("class").as_deref(),
                    x.remove("provider").as_deref(),
                    x.remove("type").as_deref(),
                ) {
                    (Some("ocf"), Some("ddn"), Some("lustre-server")) => {
                        current_ra = Some(Ra::Ddn(None));
                    }
                    (Some("ocf"), Some("lustre"), Some("Lustre")) => {
                        current_ra = Some(Ra::Lustre(None));
                    }
                    _ => {}
                }
            }
            Event::Empty(ref x) if x.name() == b"nvpair" => {
                let mut x = attrs_to_hashmap(x.attributes(), &reader)?;

                match current_ra.as_mut() {
                    Some(Ra::Ddn(ref mut v)) => {
                        if x.remove("name") == Some("directory".into()) {
                            *v = x.remove("value");
                        }
                    }
                    Some(Ra::Lustre(ref mut v)) => {
                        if x.remove("name") == Some("mountpoint".into()) {
                            *v = x.remove("value");
                        }
                    }
                    _ => {}
                }
            }

            Event::Eof => break,
            _ => {}
        };

        buf.clear();
    }

    Ok(current_ra.and_then(|x| x.inner()))
}

fn read_resources(reader: &mut Reader<&[u8]>) -> Result<Vec<Resource>, ImlAgentError> {
    let mut buf = vec![];
    let mut xs = vec![];

    let mut current_resource = None;
    let mut clone_index = -1;

    loop {
        buf.clear();

        match reader.read_event(&mut buf)? {
            Event::Start(x) => match x.name() {
                b"resource" => {
                    let x = attrs_to_hashmap(x.attributes(), reader)?;

                    current_resource = Some(resource_from_map(&x)?);
                }
                b"clone" => {
                    clone_index = 0;
                }
                _ => {}
            },
            Event::Empty(x) => match x.name() {
                b"node" => {
                    let mut x = attrs_to_hashmap(x.attributes(), reader)?;

                    if let Some(current_resource) = current_resource.as_mut() {
                        current_resource.active_node_name = x.remove("name");
                        current_resource.active_node_id = x.remove("id");
                    }
                }
                b"resource" => {
                    let x = attrs_to_hashmap(x.attributes(), reader)?;
                    let mut r = resource_from_map(&x)?;
                    if clone_index >= 0 {
                        r.id = format!("{}:{}", &r.id, clone_index);
                        clone_index += 1;
                    }

                    xs.push(r);
                }
                _ => {}
            },
            Event::End(x) => match x.name() {
                b"resource" => {
                    if let Some(mut x) = current_resource.take() {
                        if clone_index >= 0 {
                            x.id = format!("{}:{}", &x.id, clone_index);
                            clone_index += 1;
                        }
                        xs.push(x);
                    }
                }
                b"clone" => {
                    clone_index = -1;
                }
                b"resources" => {
                    break;
                }
                _ => {}
            },
            _ => {}
        }
    }

    Ok(xs)
}

fn read_bans(reader: &mut Reader<&[u8]>) -> Result<Vec<Ban>, ImlAgentError> {
    let mut buf = vec![];
    let mut xs = vec![];

    loop {
        buf.clear();

        match reader.read_event(&mut buf)? {
            Event::Empty(x) if x.name() == b"ban" => {
                let x = attrs_to_hashmap(x.attributes(), reader)?;
                let x = ban_from_map(&x)?;

                xs.push(x);
            }
            Event::End(x) if x.name() == b"bans" => {
                break;
            }
            _ => {}
        }
    }

    Ok(xs)
}

pub(crate) fn read_crm_output(crm_output: &[u8]) -> Result<Cluster, ImlAgentError> {
    let x = std::str::from_utf8(crm_output)?;

    let mut reader = Reader::from_str(x);
    reader.trim_text(true);

    let mut buf = vec![];

    let mut cluster = Cluster::default();

    loop {
        match reader.read_event(&mut buf)? {
            Event::Start(ref x) => match x.name() {
                b"nodes" => {
                    cluster.nodes = read_nodes(&mut reader)?;
                }
                b"resources" => {
                    cluster.resources = read_resources(&mut reader)?;
                }
                _ => {}
            },
            Event::Eof => break,
            _ => {}
        };

        buf.clear();
    }

    Ok(cluster)
}

fn read_banned_output(crm_output: &[u8], cluster: &mut Cluster) -> Result<(), ImlAgentError> {
    let x = std::str::from_utf8(crm_output)?;

    let mut reader = Reader::from_str(x);
    reader.trim_text(true);

    let mut buf = vec![];

    loop {
        match reader.read_event(&mut buf)? {
            Event::Start(ref x) if x.name() == b"bans" => {
                cluster.bans = read_bans(&mut reader)?;
            }
            Event::Eof => break,
            _ => {}
        };

        buf.clear();
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    static ES_FIXTURE: &'static [u8] = include_bytes!("./fixtures/es_fixture.xml");
    static ES_BAN_FIXTURE: &'static [u8] = include_bytes!("./fixtures/es_ban_fixture.xml");
    static ES_LUSTRE_RESOURCE_MOUNTS_FIXTURE: &'static [u8] =
        include_bytes!("./fixtures/es_lustre_resource_mounts_fixture.txt");
    static VAGRANT_FIXTURE: &'static [u8] = include_bytes!("./fixtures/vagrant_fixture.xml");
    static VAGRANT_STOPPED_FIXTURE: &'static [u8] =
        include_bytes!("./fixtures/vagrant_stopped_fixture.xml");
    static COROSYNC_CFGTOOL_FIXTURE: &'static [u8] =
        include_bytes!("./fixtures/corosync_cfgtool_fixture.txt");

    #[test]
    fn test_read_es() {
        let mut cluster = read_crm_output(ES_FIXTURE).unwrap();

        read_banned_output(ES_BAN_FIXTURE, &mut cluster).unwrap();

        insta::assert_debug_snapshot!(cluster);
    }

    #[test]
    fn test_read_vagrant() {
        let cluster = read_crm_output(VAGRANT_FIXTURE).unwrap();

        insta::assert_debug_snapshot!(cluster);
    }

    #[test]
    fn test_read_vagrant_stopped() {
        let cluster = read_crm_output(VAGRANT_STOPPED_FIXTURE).unwrap();

        insta::assert_debug_snapshot!(cluster);
    }

    #[test]
    fn test_lustre_resource_mounts_output() {
        let x = parse_resource_xml(ES_LUSTRE_RESOURCE_MOUNTS_FIXTURE).unwrap();

        insta::assert_debug_snapshot!(x);
    }

    #[test]
    fn test_local_nodeid_output() {
        let x = parse_local_nodeid_output(COROSYNC_CFGTOOL_FIXTURE).unwrap();

        insta::assert_debug_snapshot!(x);
    }
}
