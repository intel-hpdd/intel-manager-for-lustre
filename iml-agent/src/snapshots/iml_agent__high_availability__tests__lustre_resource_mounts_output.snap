---
source: iml-agent/src/high_availability.rs
expression: x
---
Some(
    "/lustre/fs0a9c/ost0003",
)
